ibtk/include/ibtk/SnapshotCache.h

```diff
@@ -0,0 +1,226 @@
+    }
+
+    /*!
+     * Get the set of data points at which snapshots are stored.
+     */
+    inline const std::set<double>& getSnapshotTimes() const
+    {
+        return d_snapshot_times;
+    }
+
+    /*!
+     * Write the snapshot cache to a database. This writes both the time points and the snapshotted patch hierarchies.
+     */
+    void putToDatabase(SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> db) override;
+
+private:
+    /*!
+     * Fill snapshot_idx with u_idx. This assumes that u_idx is a valid patch data index on the specified hierarchy. No
+     * error checking is done with this function. Note that the snapshot hierarchy and time value in the member map will
+     * be overwritten if currently initialized.
+     */
+    void fillSnapshot(int u_idx,
+                      SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy,
+                      double time);
+
+    /*!
+     * Read stored information from the restart files. Note the GridGeometry object is required in order to construct
+     * the snapshotted patch hierarchies.
+     */
+    void getFromRestart(SAMRAI::tbox::Pointer<SAMRAI::hier::GridGeometry<NDIM> > grid_geom);
+
+    /*!
+     * \brief Default constructor.
+     *
+     * \note This constructor is not implemented and should not be used.
+     */
+    SnapshotCache() = delete;
+
+    /*!
+     * \brief Copy constructor.
+     *
+     * \note This constructor is not implemented and should not be used.
+     *
+     * \param from The value to copy to this object.
+     */
+    SnapshotCache(const SnapshotCache& from) = delete;
+
+    /*!
+     * \brief Assignment operator.
+     *
+     * \note This operator is not implemented and should not be used.
+     *
+     * \param that The value to assign to this object.
+     *
+     * \return A reference to this object.
+     */
+    SnapshotCache& operator=(const SnapshotCache& that) = delete;
+
+    std::string d_object_name;
+
+    /*
+     * Data for tracking snapshots.
+     */
+    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > d_snapshot_var;
+    SAMRAI::tbox::Pointer<SAMRAI::hier::VariableContext> d_ctx;
+    int d_snapshot_idx = IBTK::invalid_index;
+    std::string d_snapshot_refine_type = "CONSERVATIVE_LINEAR_REFINE";
+    SAMRAI::hier::IntVector<NDIM> d_gcw = 1;
+
+    /*
+     * Vector of the time points at which the snapshots are taken.
+     */
+    std::set<double> d_snapshot_times;
+
+    /*!
+     * We store a copy of the patch hierarchy at which the snapshot is stored. As needed, we will transfer data
+     * from the snapshot hierarchy to the provided hierarchy.
+     *
+     * We use a map with a double as a key. This means only values stored in d_snapshot_times should be used to access
+     * an element in this map, otherwise we may not find an element because of finite precision.
+     */
+    std::map<double, SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > > d_snapshot_hierarchies;
```

Comment:
I don't think its a good idea to store the keys in two places at once - we might update one and not the other.

A better alternative would be to store
```cpp
std::vector<std::pair<double, tbox::Pointer<hier::PatchHierarchy>>
```
instead and keep that sorted ourselves.